### PR TITLE
[New Command] Added new function New-DbaDbUser

### LINKED ISSUE #2851: New Command: New-DbaDbUser to add a user to a database
## Feature Request

There is an existing command to Remove-DbaDbUser but no corresponding Add command.

### Problem to solve
This is an inconsistency. 

### Additional information

Accept pipeline input from New-DbaLogin
Returns the newly created user

### PR DESCRIPTION
**Pull request by frank687**
## Type of Change

 - [x] Bug fix (non-breaking change, fixes #2851 )
 - [x] New feature (non-breaking change, adds functionality)
 - [ ] Breaking change (effects multiple commands or functionality)
 - [ ] Ran manual Pester test and has passed (`.\tests\manual.pester.ps1)
 - [ ] Adding code coverage to existing functionality
 - [ ] Pester test is included
 - [ ] Nunit test is included
 - [ ] Documentation
 - [ ] Build system
 

### Purpose
 
Adds the New-DbaDbLogin function based upon the issue request.

### Commands to test

New-DbaDbUser -SqlInstance Instance1 -Database DB1 -Login Login2 
New-DbaDbUser -SqlInstance Instance1 -Database DB1 -Login Login2 -Force
New-DbaDbUser -SqlInstance Instance1 -Database DB1 -Userrname SomeUser 
New-DbaDbUser -SqlInstance Instance1 -Database DB1 -Userrname SomeUser -Force

### CODE REVIEW

**Comment by frank687** [comment]
This failed appveyor's indentation test but I dont see any leading spaces anymore.

**Comment by potatoqualitee** [comment]
thanks so much frank! i just commited some changes, hopefully that works.

@ClaudioESSilva or @wsmelton would you be able to test/review this PR? 🙇

File: functions/New-DbaDbUser.ps1 [hunk 1]
```diff
@@ -0,0 +1,247 @@
...
+		.PARAMETER Database
+			Specifies the database(s) to process. Options for this list are auto-populated from the server. If unspecified, all databases will be processed.
+		.PARAMETER ExcludeDatabase
+			Specifies the database(s) to exclude from processing. Options for this list are auto-populated from the server. By default, system databases are excluded.
+		.PARAMETER IncludeSystem
+			If this switch is enabled, the user will be added to system databases.
+		.PARAMETER Login
+			When specified, the user will be associated to this SQL login and have the same name as the Login. Username or Login may be specified but not both.
+		.PARAMETER Username
+			When specified, the user will have this name. Username or Login may be specified but not both.
```
Comment (wsmelton) [inline functions/New-DbaDbUser.ps1]: Without looking at the code, just based on help info...

Does this mean then that I can't add a Windows Login and map it to a different username within a given database? SQL Server allows the username in the database to be different than the login it is mapped to.

File: functions/New-DbaDbUser.ps1 [hunk 2]
```diff
@@ -0,0 +1,247 @@
...
+		.PARAMETER Force
+			If user exists, drop and recreate.
```
Comment (wsmelton) [inline functions/New-DbaDbUser.ps1]: What about the login?

File: functions/New-DbaDbUser.ps1 [hunk 3]
```diff
@@ -0,0 +1,247 @@
...
+        [parameter(ParameterSetName = "Login")]
+        [string[]]$Login,
+        # [parameter(ParameterSetName = "Contained")]
+
+        [parameter(ParameterSetName = "Login")]
+        [parameter(ParameterSetName = "NoLogin")]
+        [string[]]$Username,
+        # [parameter(ParameterSetName = "Contained")]
+
+        # [Security.SecureString]$Password,
```
Comment (wsmelton) [inline functions/New-DbaDbUser.ps1]: Need to remove the commented out parameter names.

File: functions/New-DbaDbUser.ps1 [hunk 4]
```diff
@@ -0,0 +1,247 @@
...
+
+        [switch]$Force,
+        [switch][Alias('Silent')]
+        $EnableException
+    )
+    
+    begin {
+        
+        
+    }
```
Comment (wsmelton) [inline functions/New-DbaDbUser.ps1]: Remove begin block if not in use.

File: functions/New-DbaDbUser.ps1 [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+            try {
+                Write-Message -Level Verbose -Message "Connecting to $instance"
+                $server = Connect-SqlInstance -SqlInstance $instance -SqlCredential $sqlcredential
+            }
+            catch {
+                Stop-Function -Message "Failure" -Category ConnectionError -ErrorRecord $_ -Target $instance -Continue
+                return
+            }
+            
+            $databases = $server.Databases | Where-Object Status -EQ "normal"
```
Comment (wsmelton) [inline functions/New-DbaDbUser.ps1]: This needs to be based on `IsAccessible`, the status of normal won't include databases that are not accessible due to read-only access or involved in mirroring and such.

File: functions/New-DbaDbUser.ps1 [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+            
+            if ($Database) {
+                $databases = $databases | Where-Object Name -In $Database
+            }
+            if ($ExcludeDatabase) {
+                $databases = $databases | Where-Object Name -NotIn $ExcludeDatabase
+            }
+            If (-Not $IncludeSystem) {
```
Comment (wsmelton) [inline functions/New-DbaDbUser.ps1]: Lowercase `if` to match standard in code.

File: functions/New-DbaDbUser.ps1 [hunk 7]
```diff
@@ -0,0 +1,247 @@
...
+                        }
+                        
+                        
+                        if ($Login.GetType().Name -eq 'Login') {
+                            $smoLogin = $Login
+                        }
+                        else {
+                            #get the login associated with the given name.
+                            $smoLogin = $server.Logins | Where-Object Name -eq $Login
+                            if ($smoLogin -eq $null) { Stop-Function -Message "Invalid Login: $Login is not found on $Server" -Target $instance; return }
```
Comment (wsmelton) [inline functions/New-DbaDbUser.ps1]: Format, this needs to be split across multiple lines for readability.

The return needs to be removed and `Stop-Function` should have the `-Continue` parameter added. By using the return you stop the call of the function completely. You provide the ability to pass in multiple logins, so the first one that is now found stops any of the other ones from being processed.

Which if multiple logins are passed in `Where-Object Name -eq $Login` will always come up empty. Without testing you likely need to provide a loop to iterate over each login passed in.

File: functions/New-DbaDbUser.ps1 [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+                    #         Stop-Function -Message "Sql User's with passwords can only be created on partially contained databases."  -Category InvalidOperation -Target $instance;
+                    #         return
+                    #     } else {
+                    #         Write-Message -Level Verbose -Message $db.ContainmentType
+                    #     }
+                    
+                    
+                    #     $UserType = [Microsoft.SqlServer.Management.Smo.UserType]::SqlUser
+                    
+                    # }
```
Comment (wsmelton) [inline functions/New-DbaDbUser.ps1]: Commented code needs to be removed

File: functions/New-DbaDbUser.ps1 [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+                        
+                        If ($PSBoundParameters.Keys -contains 'Login' -and $Login.GetType().Name -eq 'Login') { $smoUser.Login = Login }
+                        $smoUser.UserType = $UserType;
+                        
+                        $smoUser.Create();
+                        
+                    }
+                    catch {
+                        Stop-Function -Message "Failed to add user $Name in $db to $instance" -Category InvalidOperation -ErrorRecord $_ -Target $instance -Continue;
+                        return
```
Comment (wsmelton) [inline functions/New-DbaDbUser.ps1]: This is a contradiction, to have continue on `Stop-Function` but also include the `return`. Need to remove return if you want the command to continue processing after the catch block is hit.

File: functions/New-DbaDbUser.ps1 [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+                    
+                    if ($PSBoundParameters.Keys -contains 'Username' -and $smoUser.Name -ne $Username) {
+                        $smoUser.Rename($Username);
+                    }
+                    
+                    # if ( $PSBoundParameters.Keys -contains 'Password' ) {
+                    #     $smoUser.ChangePassword('',$Password);
+                    #     $smoUser.Alter();
+                    #     $smoUser.Refresh();
+                    # }
```
Comment (wsmelton) [inline functions/New-DbaDbUser.ps1]: Need to remove commented code.

File: functions/New-DbaDbUser.ps1 [hunk 11]
```diff
@@ -0,0 +1,247 @@
...
+                    
+                }
+                
+                #Display Results
+                Get-DbaDatabaseUser -SqlInstance $server.Name -Database $db.Name | Where-Object name -eq $smoUser.Name
+            } #foreach ($db in $databases)
+        } #foreach ($instance in $SqlInstance)
+    }
+    end {
+        if (Test-FunctionInterrupt) { return }
```
Comment (wsmelton) [inline functions/New-DbaDbUser.ps1]: All the returns in this command need to be removed if you are going to allow processing of multiple databases, logins, and users. Once that is done this end block can be removed as well.

**Comment by potatoqualitee** [comment]
hey @frank687 ! I don't want your PR to feel neglected. I'll be merging once the change are approved 👍

**Comment by frank687** [comment]
@potatoqualitee Not neglected, working on the issues identified and trying to develop pester tests.

**Comment by potatoqualitee** [comment]
So sorry I was unclear! It's not you - I was apologizing for me taking a few days to get around to your PR. 

I don't have the mental bandwidth for code reviews right now because of the work I'm doing for Bits, so will merge once @wsmelton gives the approval.

**Comment by potatoqualitee** [comment]
awesome, thanks 💯 💯 💯

**Comment by awickham10** [comment]
@frank687 are you still working on this? I was going through the Bill of Health and finding some functions to create tests for and randomly picked this one before I realized this thread existed. Are you still planning on submitting more changes for this? If so I'll toss my stuff.

**Comment by frank687** [comment]
not actively at the moment. I cant seem to get tests to work locally for me
anyway

**Pull request merged**